File: Java/0019-remove-nth-node-from-end-of-list.java
// Definition for singly-linked list.
public class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

public class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        // Dummy so we can handle removal of the head uniformly
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        
        ListNode fast = dummy;
        ListNode slow = dummy;
        
        // Advance fast by n+1 steps
        for (int i = 0; i <= n; i++) {
            fast = fast.next;
        }
        
        // Move both pointers until fast reaches the end
        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
        }
        
        // slow.next is the node to remove
        slow.next = slow.next.next;
        
        // Return the real head (which may have changed)
        return dummy.next;
    }
}

File: Go/0019-remove-nth-node-from-end-of-list.go
// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	// Dummy so deletions at head are uniform
	dummy := &ListNode{Next: head}
	fast, slow := dummy, dummy

	// Move fast ahead by n+1 steps
	for i := 0; i <= n; i++ {
		fast = fast.Next
	}
	// Advance both until fast reaches the end
	for fast != nil {
		fast = fast.Next
		slow = slow.Next
	}
	// slow.Next is the node to remove
	slow.Next = slow.Next.Next

	return dummy.Next
}
